### PR TITLE
Tc12 add products in cart

### PR DESCRIPTION
**Pull request by Marina-Nabil-Gohar223**
workflow : https://github.com/MahmoudElSharkawyGS/AutomationExercisePractice/actions/runs/8088420362

### CODE REVIEW

File: src/main/java/com/gizasystems/automationexercise/pages/CartPage.java [hunk 1]
```diff
@@ -13,30 +13,81 @@ public class CartPage {
...
+
+    private final By secondProductPrice = By.xpath("//td[@class='cart_price']//p[contains(text(),'Rs. 400')]");
+
+    private final By firstProductQuantity = By.xpath("//tr[@id='product-1']//button[@class='disabled'][normalize-space()='1']");
+
+    private final By secondProductQuantity = By.xpath("//tr[@id='product-2']//button[@class='disabled'][normalize-space()='1']");
+
+    private final By firstProductTotal = By.xpath("//p[@class='cart_total_price'][normalize-space()='Rs. 500']");
+
+    private final By secondProductTotal = By.xpath("//p[@class='cart_total_price'][normalize-space()='Rs. 400']");
```
Comment (MahmoudElSharkawyGS) [inline src/main/java/com/gizasystems/automationexercise/pages/CartPage.java:32]: - I believe we can enhance those locators (let's discuss this next session).
- Please remove the empty lines between the locators.

File: src/main/java/com/gizasystems/automationexercise/pages/CartPage.java [hunk 2]
```diff
@@ -13,30 +13,81 @@ public class CartPage {
...
-    public CartPage verifyProductAddedToCart(String addedProductName){
+    public CartPage verifyProductAddedToCart(String addedProductName) {
         driver.assertThat().element(productName_h4).text().isEqualTo(addedProductName).perform();
         return this;
     }
+
+    public CartPage navigate() {
+        driver.browser().navigateToURL(url);
+        return this;
+    }
```
Comment (MahmoudElSharkawyGS) [inline src/main/java/com/gizasystems/automationexercise/pages/CartPage.java:64]: navigation method is not a validation method, please move it to the top on the actions section.

File: src/main/java/com/gizasystems/automationexercise/pages/CartPage.java [hunk 3]
```diff
@@ -13,30 +13,81 @@ public class CartPage {
...
+        driver.verifyThat().element(secondProductQuantity).text().isEqualTo(secondItemQuantity).perform();
+        return this;
+    }
+
+    @Step("Validate on the Total Price of the Products")
+    public CartPage validateOnTotalPrice(String firstItemTotalPrice, String secondItemTotalPrice) {
+        driver.verifyThat().element(firstProductTotal).text().isEqualTo(firstItemTotalPrice).perform();
+        driver.verifyThat().element(secondProductTotal).text().isEqualTo(secondItemTotalPrice).perform();
+        return this;
+    }
```
Comment (MahmoudElSharkawyGS) [inline src/main/java/com/gizasystems/automationexercise/pages/CartPage.java:92]: Those methods are not dynamic; What if we have only 1 item, or if we have more than 2 items?.
We need to make those methods dynamic, not only works with 2 items.

File: src/main/java/com/gizasystems/automationexercise/pages/NavigationBar.java [hunk 4]
```diff
@@ -13,6 +13,9 @@ public class NavigationBar {
     private final By delete_link = By.cssSelector("a[href='/delete_account']");
     private final By loggedInUser_link = By.xpath("//i[@class = 'fa fa-user']/parent::a");
```
Comment (MahmoudElSharkawyGS) [inline src/main/java/com/gizasystems/automationexercise/pages/NavigationBar.java:15]: Delete this empty line here between the locators please.

File: src/main/java/com/gizasystems/automationexercise/pages/NavigationBar.java [hunk 5]
```diff
@@ -13,6 +13,9 @@ public class NavigationBar {
...
 
+    private final By product_link = By.cssSelector("a[href='/products']");
+
+
```
Comment (MahmoudElSharkawyGS) [inline src/main/java/com/gizasystems/automationexercise/pages/NavigationBar.java:18]: Delete one empty line of the two here please.

File: src/main/java/com/gizasystems/automationexercise/pages/ProductsPage.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    private String url = System.getProperty("baseUrl") + "/products";
+
+    private final By HoverOnItemOne = By.xpath("(//div[@class=\"single-products\"]//img)[1]");
+    private final By ClickOnItemOne = By.xpath("(//a[@data-product-id=\"1\"])[1]");
+
+    private final By HoverOnItemTwo = By.xpath("(//div[@class=\"single-products\"]//img)[2]");
+    private final By ClickOnItemTwo = By.xpath("(//a[@data-product-id=\"2\"])[2]");
+
+    private final By continueButton = By.xpath("(//button[@class=\"btn btn-success close-modal btn-block\"])[1]");
+    private final By viewCartButton = By.xpath("//u[normalize-space()='View Cart']");
```
Comment (MahmoudElSharkawyGS) [inline src/main/java/com/gizasystems/automationexercise/pages/ProductsPage.java:21]: Locators names should be descripting what they are, not what they do (you can do more than one action on the same locator).
Also we can remove the empty line between them.
One more thing, I believe we can enhance those locators.

File: src/main/java/com/gizasystems/automationexercise/pages/ProductsPage.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        return this;
+    }
+
+    @Step("Add Second Product to Cart")
+    public ProductsPage addSecondProductToCart() {
+        driver.element().hover(HoverOnItemTwo);
+        driver.element().click(ClickOnItemTwo);
+        driver.element().click(viewCartButton);
+        return this;
+    }
```
Comment (MahmoudElSharkawyGS) [inline src/main/java/com/gizasystems/automationexercise/pages/ProductsPage.java:46]: Both methods are identical, we need to make one dynamic method.

**Review by MahmoudElSharkawyGS — CHANGES_REQUESTED** [review]
Please make sure to execute the pipeline again after the comments changes.

**Comment by MahmoudElSharkawyGS** [comment]
@Marina-Nabil-Gohar223 Any updates regarding tis PR please?

**Comment by Marina-Nabil-Gohar223** [comment]
> @Marina-Nabil-Gohar223 Any updates regarding tis PR please?


Yes, there is an update on the pull request

**Review by MahmoudElSharkawyGS — CHANGES_REQUESTED** [review]
- You are using alot of dynamic locators which sometime can complicate the cases; Let's discuss this please.
- Also, please make sure to delete any empty lines as agreed before.
- Please make sure to resolve any conflicts before pushing.
- Finally, please make sure to execute the pipeline again and every change and share the execution pipeline link.

File: src/main/java/com/gizasystems/automationexercise/pages/CartPage.java [hunk 8]
```diff
@@ -16,17 +19,35 @@ public class CartPage {
...
+    private By productPrice(Integer ID) {
+        return By.xpath("//tr[@id=\"product-" + ID + "\"]//td[@class='cart_price']//p");
+    }
+    private By productQuantity(Integer ID) {
+        return By.xpath("//tr[@id=\"product-" + ID + "\"]//button[@class='disabled']");
+    }
+    private By productTotalPrice(Integer ID) {
+        return By.xpath("//tr[@id=\"product-" + ID + "\"]//p[@class='cart_total_price']");
+    }
```
Comment (MahmoudElSharkawyGS) [inline src/main/java/com/gizasystems/automationexercise/pages/CartPage.java:33]: - Not sure to accept those locators as the user wouldn't know the ids of the items, this should be something like the name of the item, not the id
- Please follow the naming convention for the parameters variables (lowerCamaelCase).. ID should be id

File: src/main/java/com/gizasystems/automationexercise/pages/ProductsPage.java [hunk 9]
```diff
@@ -22,6 +24,15 @@ private By viewProduct_link(String productName) {
...
     private final By searchResult = By.xpath("//div[@class='productinfo text-center']//p");
 
+    private By hoverOnProduct(Integer Index) {
+        return By.xpath("(//div[@class=\"single-products\"]//img)[" + Index + "]");
+    }
+
+    private By ClickOnProduct(Integer Index) {
+        return By.xpath("(//a[@data-product-id=\"" + Index + "\"])[" + Index + "]");
+
+    }
```
Comment (MahmoudElSharkawyGS) [inline src/main/java/com/gizasystems/automationexercise/pages/ProductsPage.java:34]: - Not sure to accept those locators as the user wouldn't know the index of the items if the order changes, this should be something like the name of the item, not the index
- Please follow the naming convention for the parameters variables (lowerCamaelCase).. Index should be index.

File: src/test/java/com/gizasystems/automationexercise/tests/shoppingCart/AddProductInCart.java [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+        new CartPage(driver)
+                .navigate()
+                .validateOnItemsAddedInCart(testData.getTestData("FirstProduct.productDescription"))
+                .validateOnItemsAddedInCart(testData.getTestData("SecondProduct.productDescription"))
+                .validateOnProductPrices(1, testData.getTestData("FirstProduct.productPrice"))
+                .validateOnProductPrices(2, testData.getTestData("SecondProduct.productPrice"))
+                .validateOnProductQuantity(1, testData.getTestData("FirstProduct.productQuantity"))
+                .validateOnProductQuantity(2, testData.getTestData("SecondProduct.productQuantity"))
+                .validateOnTotalPrice(1, testData.getTestData("FirstProduct.total"))
+                .validateOnTotalPrice(2, testData.getTestData("SecondProduct.total"));
```
Comment (MahmoudElSharkawyGS) [inline src/test/java/com/gizasystems/automationexercise/tests/shoppingCart/AddProductInCart.java:41]: any data should be retrieved from the test data json file (we can't add index like this if we should put it :D)

**Review by MahmoudElSharkawyGS — CHANGES_REQUESTED** [review]

**Review by MahmoudElSharkawyGS — APPROVED** [review]
Thank you!

**Pull request merged**